tools/src/launcher/Main.hs

```diff
@@ -178,16 +183,82 @@ reportErrorDefault filename contents = do
     createDirectoryIfMissing True logDir
     writeFile (logDir </> filename) contents
 
+data Chunk
+  = EnvVar Text
+  | Plain  Text
+  deriving (Show)
+
+parseEnvrefs :: Text -> P.Result [Chunk]
+parseEnvrefs text = P.parseString pEnvrefs mempty (toString text)
+  where
+    pEnvrefs :: (Monad p, P.TokenParsing p) => p [Chunk]
+    pEnvrefs = P.some $ P.choice [pPassiveText, pEnvvarRef]
+
+    pEnvvarRef, pPassiveText :: (Monad p, P.TokenParsing p) => p Chunk
+    -- TODO: figure out a less ugly OS conditionalisation method
+    -- TODO: decide if we want escaping
+#ifdef mingw32_HOST_OS
+    pEnvvarRef   = P.between (P.char '%') (P.char '%') pEnvvarName <&> EnvVar . toText
+    pPassiveText = P.some (P.noneOf "%")                           <&> Plain  . toText
+#else
+    pEnvvarRef   = P.char '$' >>
+                   P.choice [ pEnvvarName
+                            , P.between (P.char '{') (P.char '}') pEnvvarName
+                            ] <&> EnvVar
+    pPassiveText = P.some (P.noneOf "$") <&> Plain  . toText
+#endif
+    pEnvvarName :: (Monad p, P.TokenParsing p) => p Text
+    pEnvvarName = (P.some $ P.choice [P.alphaNum, P.char '_']) <&> toText
+
+substituteEnvVarsText :: Text -> Text -> IO Text
```

Comment:
It'd be preferable to pass the whole env in here and make it pure, rather than using IO and using `lookupEnv`. Instead call `getEnvironment` once on the outside, and use `Map.fromList` to turn it into a map, and pass the map into `substituteEnvVarsText`